Adding functions to Like model

Shalevro2 opened the pull request:
Updating models.py- adding functions to Like model

Creating test_like.py- adding tests to Like model

Resolve #29

Shalevro2 committed:
Adding functions to Like model

Updating models.py- adding functions to Like model

Creating test_like.py- adding tests to Like model

Anas-mhameed reviewed (approved):
LGTM

rawadabu reviewed (approved):
LGTM

Yarboa commented on projboard/tests/test_like.py:
```diff
@@ -0,0 +1,50 @@
+        result = Like.delete_like(like.user_id, like.article_id)
+        assert result
+
+    def test_amount_of_likes_article_1(self, generate_like):
```
Consider Querying the amount of like per article through DB
To see the amount has increased

avi-biton replied:
+1

Shalevro2 replied:
I use @pytest.mark.django_db at the head of the class, its not enough?

Shalevro2 replied:
Do u mean something like that, `len(Like.objects.filter(article_id=article_id)) == amount_likes` ?

Yarboa replied:
generate_like is a fixture 
It should create and save the data in DB
return USER_TEST, ARTICLE_TEST

Like already creted the link
The tests should retrieve number of likes based on USER and ARTICLE you already holding in fixture data

Yarboa commented on projboard/tests/test_like.py:
```diff
@@ -0,0 +1,50 @@
+ARTICLE_TEST = Article(user_id=A[0], title=A[1], subject_id=A[2], content=A[3])
+
+
+@pytest.fixture
```
Please use 4 different fixtures,
 
create_user
create_test
create article

generate like will call this 4 fixtures
@pytest.fixture
def create_user(db):
While using the db fixture
https://pytest-django.readthedocs.io/en/latest/helpers.html#fixtures

test_delete_like(create_user, create_subject..):

Shalevro2 replied:
create_test?
Do you mean that?:
create_user
create_subject
create_like
create article

rawadabu replied:
@Yarboa Do you mean creating a separate file in tests/ file for generating testing objects?

```
import pytest
from projboard.models import User, Subject, Article, Like, View


# A SEPARATE FILE FOR GENERATING TESTING OBJECTS
@pytest.fixture
def generate_user(email='user@gmail.com', password='password', name='name', nickname='nickname'):
    user = User(email=email, password=password, name=name, nickname=nickname)
    user.save()
    return user


@pytest.fixture
def generate_subject():
    return Subject


@pytest.fixture
def generate_article():
    return Article


@pytest.fixture
def generate_like():
    return Like


@pytest.fixture
def generate_view():
    return View

```

Yarboa replied:
Yes please consult @Anas-mhameed

Yarboa commented on projboard/tests/test_like.py:
```diff
@@ -0,0 +1,50 @@
+
+
+@pytest.fixture
+def generate_like():
```
see previous comment

avi-biton commented on projboard/tests/test_like.py:
```diff
@@ -0,0 +1,50 @@
+import pytest
+from projboard.models import Like, User, Article, Subject
+
+U = ['usertest@gmail.co.il', '123456', 'John Doe', 'USER_TEST']
```
Try to name the parameters with more meaningful names:
like 'user' instead of 'U'  
'article' instead of 'A'

Shalevro2 committed:
Fixing test_like.py

Changing parameters names

Adding fixtures

Changing the tests functions: test_amount_of_likes_article_0-1

finziamit commented on projboard/tests/test_like.py:
```diff
@@ -1,50 +1,58 @@
     ARTICLE_TEST.save()
-    Like(user_id=USER_TEST, article_id=ARTICLE_TEST).save()
-    return USER_TEST, ARTICLE_TEST
+    return ARTICLE_TEST
```
I'm not sure it's possible to use `.save()` without a db linked, using `@pytest.mark.django_db` or db passed in the argument `create_article(db)` should solve it (if so, the `create_article(db)` should be the right use since you would use the returning value of the function).

Shalevro2 replied:
Done at [0fd9b0c](https://github.com/redhat-beyond/Smarticle/pull/30/commits/0fd9b0c231b73f91ed35d32628a683922c15613c)

Shalevro2 committed:
Add db to fixtures

finziamit reviewed (approved):
LGTM

Shalevro2 committed:
Fixing fixtures

Adding __str__ function to like model class at models.py

Updating test_like.py- fixing the fixtures

Yarboa commented on projboard/tests/test_like.py:
```diff
@@ -0,0 +1,67 @@
+
+    @pytest.fixture
+    @pytest.mark.django_db
+    def create_user(self):
```
Please use for loop here 
You could use 
this dictionary and use loops 

USER_LIST=[{'name':'test1_name', 'email':'tt@ttt.com'},{'name':'test2_name', 'email':'tt2@ttt.com'}]

user_list=[]
for i, user in enumerate(USER_LIST)
   user= User(name=,USER_LIST[i].name nickname=USER_LIST[i].nick_name, email=..., password=....)
   user.save()
   user_list.append(user)
   
  return user_list 
  
  
  You can also receive ideas from here 
  https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#parametrizing-fixtures

Shalevro2 replied:
Hi @Yarboa,
Please take a look at [033bfc9](https://github.com/redhat-beyond/Smarticle/pull/30/commits/033bfc9d62066b95448471aa9a9d1797d788b9ec).
I Changed the fixtures and a some tests

Yarboa commented on projboard/tests/test_like.py:
```diff
@@ -0,0 +1,67 @@
+    @pytest.mark.django_db
+    def create_articles(self, create_user, create_subjects):
+        articles_list = []
+        for i in range(len(create_user)):
```
Please refer the above comment use parameters

Shalevro2 replied:
Acutely I don't need many articles, I did not use them.
I will fix that

Yarboa commented on projboard/tests/test_like.py:
```diff
@@ -0,0 +1,67 @@
+        assert create_like.article_id.title == TITLE
+
+    def test_delete_like(self, create_user, create_like):
+        result = Like.delete_like(create_user[1], create_like.article_id)
```
Please see above
avoid usage [0] or [1] -> use [i] or a class based on fixtures calls

Shalevro2 replied:
Hi @Yarboa,
Please take a look at [033bfc9](https://github.com/redhat-beyond/Smarticle/pull/30/commits/033bfc9d62066b95448471aa9a9d1797d788b9ec).
I Changed the fixtures and a some tests

Shalevro2 committed:
Changing in fixtures

Changing the fixtures in test_like.py
Adding more tests

Shalevro2 committed:
Organizing the models

Adding models directory

Inserting and creating models files- article.py, user.py, subject.py

Updating the imports in the migrations files

rawadabu reviewed (approved):
LGTM

Yarboa commented:
Please rebase main and push --force

Shalevro2 commented:
> Please rebase main and push --force

Done

Shalevro2 committed:
Fixing conflicts in main

Yarboa reviewed (approved):
Nice work
LGTM
